18F/identity-idp
# LG-7012 LG-7010 Add/Integrate new DDP Proofer

**Why**: To add several lexis nexis fraud based indicators via products such as fraudpoint, EmailAge, device fingerprinting, etc. and eventually have the DDP proofer be a drop in replacement for RDP proofer used at the InstantVerify step (so we would not have to make 2 api calls).  Initially this call will be made before InstantVerify and the RDP instantverify step will be left alone.  Note that DDP and threatmetrix are sometimes used synonymously.  RDP is the legacy platform that runs instant verify, phone finder, true id.  DDP is their new platform where we will call threatmetrix.
**How**: Use the proofing framework provided in `Proofing::Base` and `Proofing::LexisNexis:Proofer`.  The proofing call will call all the relevant products behind the scenes via one API call in DDP using the Session Query API https://portal.threatmetrix.us/kb/implementation/apis/session_query_api.htm . It will piggyback on the existing async  `ResolutionProofingJob` adding new params: `user_id` (to get email), and `threatmetrix_session_id` from #6694 for device profiling. Add a feature flag `lexisnexis_threatmetrix_enabled` to call threatmetrix before instant verify.  Note the new fields added to resolution proofing will be: `threatmetrix_session_id` (a unique session identifier created with the front end added LexisNexis javascript for device profiling, etc.), `email`, and the PII packet now uses `drivers_license_number` (like the previous AAMVA document check which validates the actual drivers license info with the state if the state supports it).  We dup the instant verify PII packet and add in the extra PII values and package it up for the call to DDP.  All this info allows lexis to formulate a risk score.  The job will continue to instant verify regardless of if the threatmetrix results or score and further on in the flow the user will face a sad screen upon fraud detection.

**How to test this locally:**

Update application.yml to:
  ```
development:
  lexisnexis_threatmetrix_api_key: XXXX <--- fill these in
  lexisnexis_threatmetrix_base_url: XXXX
  lexisnexis_threatmetrix_org_id: XXXX
  lexisnexis_threatmetrix_enabled: true
  proofing_device_profiling_collecting_enabled: true
```

The proofers: document (trueid), resolution (instant verify), address (phone finder)  are mocked by default (proofer_mock_fallback: true) so you can go through the verification process with bogus data (images,etc.).  But we want the threatmetrix live locally so just hardcode out that mock on the lexisnexis_ddp_proofer memoization line:
```
def lexisnexis_ddp_proofer
    @lexisnexis_ddp_proofer ||=
      if false # IdentityConfig.store.proofer_mock_fallback <- forcibly skip mock
```

Probably unwise to add in code to make it run locally (and override the global mock flag) with another feature flag as we might accidentally set one live in the lower environments where mocking is necessary.

Create an account and on the account screen type in the path `/verify` and verify your identity up to the confirm information screen. 

 After submitting there you should see an entry in (async jobs) workers.log:
```
{"name":"ThreatMetrix","user_id":"f36cef72-199c-48c1-9c1c-6bf915d29124","threatmetrix_request_id":"ddp-mock-transaction-id-123","threatmetrix_success":true}
```
Note: request_id can be used with LN to track individual calls.  In costing it's added to transaction_id.  On Lexis legacy system this was conversation_id.

Remember that not all test data in idv is equal.  You can feed it specific scenarios to generate pass/fail conditions noted here: https://developers.login.gov/testing/

Assumptions: a decision was made not to send phone even though we have it with users that have a phone mfa already configured.  

Notes: username, password are part of the legacy rdp that used basic auth.  the new ddp uses api keys and org id.  Also workflow modified the path associated with the request in the legacy rdp system (instant verify, etc.).  We are calling one endpoint here so no workflow is required.
